### PR TITLE
Add initial code for sync to async conversion

### PR DESCRIPTION
**Pull request by WillBrock**
Expanding off of [this pull request](https://github.com/webdriverio/codemod/pull/46)

This looks to do the sync to async conversion on most things.

This should convert:
- Transforms `describe`, `before`, `it` to async
- Page class method calls, including chained calls
- Sets all method definitions to async, excludes getters and setters.
- All built in methods to async
- By default, this adds awaits to most everything but the small amount of calls that don't need to be awaited can be added to the constants file.

### Add to constants arrays for custom items unique to a codebase

In `async/constants.js`

`AWAIT_CUSTOM_GETTERS` - Add getters here that will need to be awaited. e.g. things like `$$`
```
get foo() { $$(`.foobar`); }
```

`EXCLUDE_METHODS` - Add class methods here that don't need to be async / await

### CODE REVIEW

**Comment by WillBrock** [comment]
For the `forEach` parts I think it might be easiest to do the following with jscodeshift but I'm not sure how to do it.

From:
```
[1,2,3].forEach(num => {
     browser.pause(5000);
})
```

To:
```
await Promise.all([1,2,3].map(async num => {
    await browser.pause(5000);
});
```

Any suggestions would be helpful.

**Comment by christian-bromann** [comment]
@WillBrock thanks for the work! Have you tried https://astexplorer.net/? It is really great to understand how one piece of code needs to be replaced with the other. Also would you mind extending the test cases so we are sure to not introduce any bugs down the line?

**Comment by WillBrock** [comment]
Fixed a few things and added converting `forEach` to regular `for` loops. Still a work in progress.

Will add some tests as well.

**Comment by WillBrock** [comment]
I think I've got the majority of things covered. I added tests for it but the tsx parser ones are failing because it doesn't add async to method definitions. I'm not sure why those are failing so something I'll need to read up more on.

File: test/__fixtures__/async/transformed/spec.js [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        await browser.execute(() => {
+            document.querySelector('.foo').click();
+            $('.foo').click();
+        });
+    })
+
+    it('testing loops', async () => {
+        for (const num of [1, 2, 3]) {
+            await $('.foo').setValue(num)
+        }[(1, 2, 3)].map(async num => {
```
Comment (WillBrock) [inline test/__fixtures__/async/transformed/spec.js:30]: I found an issue here I need to look into.

Comment (WillBrock) [reply test/__fixtures__/async/transformed/spec.js:30]: `[(1, 2, 3)].map`

That is automatically being added by the parser when there are no transforms applied and the expression above it doesn't have an ending semi colon.

**Comment by WillBrock** [comment]
Still some more edge cases I need to do that are coming up.

**Comment by christian-bromann** [comment]
@WillBrock great work! Let me know when this is ready to review and I am happy to take a look.

**Comment by kailin0512** [comment]
@WillBrock Great work! is it ready to use or you need some help on that?

**Comment by WillBrock** [comment]
@kailin0512 It's just about ready. I've just got a few things to clean up and I'll try to push the changes tomorrow.

**Comment by WillBrock** [comment]
I think this is basically ready. I'm still going through some of our codebase at work (about 3k files) but most things are working.

If anyone wants to try it out feel free and let me know what issues you run into.

**Comment by christian-bromann** [comment]
Awesome work @WillBrock , let me review it asap.

**Comment by owens-ben** [comment]
@christian-bromann any update on approving this?

**Comment by jameskip** [comment]
Thank you @WillBrock for your work on this!

**Pull request merged**